scripts/emulator-testing/emulators/emulator.ts

```diff
@@ -53,64 +53,78 @@ export abstract class Emulator {
       return Promise.resolve();
     }
 
+    console.log(`Downloading emulator from [${this.binaryUrl}] ...`);
+    const { name: tempDir } = tmp.dirSync({ unsafeCleanup: true });
+    const filepath = path.resolve(tempDir, this.binaryName);
+    let cur = 0;
+    // The emulators vary in size and are approximately 32MiB. If we define this array to be
+    // only 32MiB, there's a risk in the future that the download will suddenly fail because there
+    // was an increase in emulator size, and the buffer is no longer large enough.
+    // To prevent this, we give some room for an increase in size of the emulators
+    // and make the buffer 64MiB.
+    let buf = new Uint8Array(2 ** 26); // 64 MiB
     return new Promise<void>((resolve, reject) => {
-      tmp.dir((err: Error | null, dir: string) => {
-        if (err) reject(err);
-        console.log(`Created temporary directory at [${dir}].`);
-        const filepath: string = path.resolve(dir, this.binaryName);
-        const writer = fs.createWriteStream(filepath);
-        console.log(`Downloading emulator from [${this.binaryUrl}] ...`);
-        // Map the DOM's fetch Reader to node's streaming file system
-        // operations. We will need to access class members `binaryPath` and `copyToCache` after the
-        // download completes. It's a compilation error to pass `this` into the named function
-        // `readChunk`, so the download operation is wrapped in a promise that we wait upon.
-        const downloadPromise = new Promise<void>(
-          (downloadComplete, downloadFailed) => {
-            fetch(this.binaryUrl)
-              .then(resp => {
-                if (resp.status !== 200 || resp.body === null) {
-                  console.log('Download of emulator failed: ', resp.statusText);
-                  downloadFailed();
-                } else {
-                  const reader = resp.body.getReader();
-                  reader.read().then(function readChunk({ done, value }): any {
-                    if (done) {
-                      downloadComplete();
-                    } else {
-                      writer.write(value);
-                      return reader.read().then(readChunk);
-                    }
-                  });
+      /**
+       * Once the download is `done` in `readChunk`, we want to set `this.binaryPath` to the path of the
+       * downloaded emulator. Unfortunately, we can't access `this` when inside `readChunk`'s scope, since
+       * it's a named function expression, and does not inherit the `this` object from it's parent.
+       * To work around this, we wrap the fetch in a promise,
+       * then once it's resolved we can access `this` in a callback arrow function that *does* inherit
```

Comment:
There's a couple of other ways around this, I think, not 100% sure. One is to set like `const self = this` at the top level (outside of the fetch anyway) and then you call `self.binaryPath` when needed. That looks weird so another better looking option is that you create a function at the top level called `setBinaryPath` that is just like

```
const setBinaryPath = (binaryPath) => this.binaryPath = binaryPath;
```
And then I think you can call setBinaryPath inside readChunk. I'm not 100% sure but I think they should work? If they don't, never mind.